Trims the buildbox

tcsc opened the pull request:
Enables the docker BuildKit in an attempt to speed up builds, and trims just under 2GB off the image size.

Prior to these changes, `docker image ls` reported the image as 6.08GB after a local build, and ~~5.06GB~~ 4.17GB after

See-Also: #6149

tcsc committed:
Trims the buildbox image and uses buildkit

 - Enables the docker BuildKit in an attempt to speed up builds
 - Trims ~1GB off image size

tcsc committed:
Merge branch 'master' into tcsc/trim-buildbox-buildkit

webvictim commented on build.assets/Dockerfile:
```diff
@@ -107,10 +110,12 @@ RUN apt-get update -y --fix-missing && \
 ENV PATH="$PATH:/opt/google-cloud-sdk/bin"
 # beta component is required by firestore emulator: https://cloud.google.com/sdk/gcloud/reference/beta/emulators/firestore/start
-RUN gcloud components install cloud-firestore-emulator beta
+RUN gcloud components install cloud-firestore-emulator beta && \
```
Maybe set the PATH up front and condense these two steps into one, so we only remove the cache once?

tcsc replied:
Done.

webvictim commented on build.assets/Dockerfile:
```diff
@@ -172,11 +178,11 @@ RUN (export PROTOC_TARBALL=protoc-${PROTOC_VER}-linux-$(if [ "$BUILDARCH" = "amd
      chmod -R a+xr /usr/local/bin/protoc && \
-     rm /tmp/${PROTOC_TARBALL})
-RUN (git clone https://github.com/gogo/protobuf.git ${GOPATH}/src/github.com/gogo/protobuf && \
+     rm /tmp/${PROTOC_TARBALL})     
```
Random whitespace added here

tcsc replied:
Removed

zmb3 reviewed (approved):
LGTM with Gus's suggestions addressed.

r0mant reviewed (commented):
Yep, we need to do this for Centos 7 dockerfile and makefile command also. We're going to use Centos 7 for all amd64 artifacts for Teleport 10.

tcsc committed:
Dockerfile refactoring

 - Break more dependencies out into separate build stages
 - Add some simple supply-chain protections for dependencies sourced
   via git. The Docker build now checks that the commit SHAs are what
   we expect, and not just assume that the tags haven't changed.
 - Moves the `cbindgen` build to a stage to avoid pulling in extra
   dependencies not needed for the Teleport build
 - Combines the `gcloud` and firestore emulator install into one step to
   reduce the layer count.
 - Ports some of the above the Centos7 Dockerfile.

tcsc committed:
Whitespace removal

tcsc committed:
Merge remote-tracking branch 'origin/master' into tcsc/trim-buildbox-buildkit

tcsc committed:
Merge remote-tracking branch 'origin/tcsc/trim-buildbox-buildkit' into tcsc/trim-buildbox-buildkit

tcsc committed:
Merge branch 'master' into tcsc/trim-buildbox-buildkit

webvictim commented on build.assets/Makefile:
```diff
@@ -121,7 +121,7 @@ build-binaries-fips: buildbox-fips
 	if [[ "$(BUILDBOX_NAME)" == "$(BUILDBOX)" ]]; then \
 		if [[ $${DRONE} == "true" ]] && ! docker inspect --type=image $(BUILDBOX) 2>&1 >/dev/null; then docker pull $(BUILDBOX) || true; fi; \
-		docker build --platform=linux/$(RUNTIME_ARCH) \
+		 DOCKER_BUILDKIT=1 docker build --platform=linux/$(RUNTIME_ARCH) \
```
There’s also a leading space here; doesn’t make any difference to the way it runs, just can’t unsee it 😁

tcsc committed:
Gus appeasement :-)

tcsc committed:
Merge branch 'master' into tcsc/trim-buildbox-buildkit

tcsc committed:
Merge branch 'master' into tcsc/trim-buildbox-buildkit

tcsc committed:
Merge branch 'master' into tcsc/trim-buildbox-buildkit
